src/components/ec/cuda/kernel/ec_cuda_reduce_ops.h

```diff
@@ -60,165 +59,101 @@ cuFloatComplex operator* (const cuFloatComplex & first,
-        ucc_eee_task_reduce_strided_t task, uint16_t flags)                     \
-    {                                                                           \
-        ucc_reduce_cuda_strided_##NAME<_Type, _AlphaType, triggered, UNROLL>(   \
-            task, flags);                                                       \
+#define CUDA_REDUCE_WITH_OP(NAME, _OP)                                           \
+    template <typename _Type, typename _AlphaType, bool triggered, int UNROLL,   \
+              typename _TaskType>                                                \
+    __device__ void ucc_reduce_cuda_##NAME(_TaskType task, uint16_t flags)       \
+    {                                                                            \
+        _Type *        d     = (_Type *)task.dst;                                \
+        const size_t   count = task.count;                                       \
+        constexpr bool strided =                                                 \
+            std::is_same<_TaskType, ucc_eee_task_reduce_strided_t>::value;       \
+        constexpr int MAXSRCS =                                                  \
+            strided ? USHRT_MAX : UCC_EE_EXECUTOR_NUM_BUFS;                      \
+        constexpr int       ALLOC_SIZE = strided ? 1 : UCC_EE_EXECUTOR_NUM_BUFS; \
+        _Type *             s[ALLOC_SIZE];                                       \
+        _Type *             s1;                                                  \
+        _Type *             s2;                                                  \
+        __shared__ uint16_t n_src2;                                              \
+        size_t              ld;                                                  \
+        size_t              i, j, line;                                          \
+        if constexpr (strided) {                                                 \
+            n_src2 = task.n_src2;                                                \
+            s1     = (_Type *)task.src1;                                         \
+            s2     = (_Type *)task.src2;                                         \
+            ld     = task.stride / sizeof(_Type);                                \
+            ucc_assert_system(task.stride % sizeof(_Type) == 0);                 \
+        } else {                                                                 \
+            memcpy(s, task.srcs, UCC_EE_EXECUTOR_NUM_BUFS * sizeof(_Type *));    \
+            n_src2 = task.n_srcs - 1;                                            \
+            s1     = s[0];                                                       \
+        }                                                                        \
+        CUDA_REDUCE_WITH_OP_CHUNK(0, UNROLL, WARP_SIZE, _OP);                    \
+        CUDA_REDUCE_WITH_OP_CHUNK(                                               \
```

Comment:
perhaps add comment between the two calls of CUDA_REDUCE_WITH_OP_CHUNK:
 /* second call for data remainder */ 
 
 Also, is there always remainder? should second CUDA_REDUCE_WITH_OP_CHUNK always be called?

Reply:
Ok I add a comment.
There is not always a remainder (no remainder if `count` is a multiple of `UNROLL*WARP_SIZE`). However, there is no harm in calling `CUDA_REDUCE_WITH_OP_CHUNK` two times anyway because if there is no remainder we will just not enter the for loop.

Reply:
Ok, possibly just a matter of personal taste, but you can precompute offset before second call and only do second call if offset == 0. 
Not sure which way is more "costly" - adding an extra if branch or adding an extra call to the stack.